### PR TITLE
Ready: Docs for sub-namespaces. Removed REST interfaces.

### PR DESCRIPTION
**Pull request by alexmoore**
Docs for Comms, Config, Exceptions, Extensions, Messages, Models, and Util namespaces.
Moved everything else to internal where possible.

Also removed last vestiges of the REST interfaces.

### CODE REVIEW

File: src/RiakClient/Config/RiakNodeConfiguration.cs [hunk 1]
```diff
@@ -48,20 +56,31 @@ public int PbcPort
             set { this["pbcPort"] = value; }
         }
 
+        /// <inheritdoc/>
+        /// <remarks>
+        /// Can be "HTTP" or "HTTPS".
+        /// Legacy, not used anymore.
+        /// </remarks>
         [ConfigurationProperty("restScheme", DefaultValue = "http", IsRequired = false)]
+        [Obsolete("The Rest properties are no longer used.")]
```
Comment (lukebakken) [inline src/RiakClient/Config/RiakNodeConfiguration.cs]: Remove it!

File: src/RiakClient/Config/IRiakNodeConfiguration.cs [hunk 2]
```diff
@@ -19,24 +19,60 @@
...
 
+        /// <summary>
+        /// The port to use when connecting to the Protocol Buffers API.
+        /// </summary>
         int PbcPort { get; }
 
+        /// <summary>
+        /// The Scheme to use when connecting to the HTTP API.
+        /// </summary>
+        [Obsolete("The Rest properties are no longer used.")]
```
Comment (lukebakken) [inline src/RiakClient/Config/IRiakNodeConfiguration.cs]: Remove

File: src/RiakClient/Config/IRiakNodeConfiguration.cs [hunk 3]
```diff
@@ -19,24 +19,60 @@
...
         string RestScheme { get; }
 
+        /// <summary>
+        /// The port to use when connecting to the HTTP API.
+        /// </summary>
+        [Obsolete("The Rest properties are no longer used.")]
```
Comment (lukebakken) [inline src/RiakClient/Config/IRiakNodeConfiguration.cs]: Lather Rinse Repeat!

**Comment by alexmoore** [comment]
Fixed some Sandcastle errors / warnings.  The remainder warnings should be fine for now, they'll fall into 1 of 3 categories:

1. Anything with `Unknown reference link target 'T:Newtonsoft.Json.JsonWriter'.` is just complaining about not having JsonWriter docs.  This is intended.
2. Any `ShowMissingComponent` warnings in regards to the `RiakClient.Comms.IRiakConnection` class, or the `RiakClient.Messages` namespace is intended.  These are public classes that we should eventually make internal, but can't due to current designs.
3. The remainder (listed below) should be `ShowMissingComponent` warnings for protected constructors/methods/properties/etc without docs.  These aren't needed at this time, as I can't see anybody overriding those classes.  In the future we should probably seal them. 

Protected ShowMissingComponent List:
```
ShowMissingComponent: [F:RiakClient.Models.Index.SecondaryIndex`2.Container] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.Config.RiakNodeConfigurationCollection.CreateNewElement] Missing <returns> documentation 
ShowMissingComponent: [M:RiakClient.Config.RiakNodeConfigurationCollection.CreateNewElement] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.Config.RiakNodeConfigurationCollection.GetElementKey(System.Configuration.ConfigurationElement)] Missing <param name="element"/> documentation 
ShowMissingComponent: [M:RiakClient.Config.RiakNodeConfigurationCollection.GetElementKey(System.Configuration.ConfigurationElement)] Missing <returns> documentation 
ShowMissingComponent: [M:RiakClient.Config.RiakNodeConfigurationCollection.GetElementKey(System.Configuration.ConfigurationElement)] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.Exceptions.RiakException.#ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext)] Missing <param name="context"/> documentation 
ShowMissingComponent: [M:RiakClient.Exceptions.RiakException.#ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext)] Missing <param name="info"/> documentation 
ShowMissingComponent: [M:RiakClient.Exceptions.RiakInvalidDataException.#ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext)] Missing <param name="context"/> documentation 
ShowMissingComponent: [M:RiakClient.Exceptions.RiakInvalidDataException.#ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext)] Missing <param name="info"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.Index.SecondaryIndex`2.#ctor(RiakClient.Models.RiakObject,System.String)] Missing <param name="container"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.Index.SecondaryIndex`2.#ctor(RiakClient.Models.RiakObject,System.String)] Missing <param name="name"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakIndexInput.#ctor(RiakClient.Models.RiakIndexId)] Missing <param name="indexId"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakIndexInput.WriteIndexHeaderJson(Newtonsoft.Json.JsonWriter)] Missing <param name="writer"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakIndexInput.WriteIndexHeaderJson(Newtonsoft.Json.JsonWriter)] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakPhaseInput.WriteBucketKeyBucketJson(Newtonsoft.Json.JsonWriter,System.String,System.String)] Missing <param name="bucketName"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakPhaseInput.WriteBucketKeyBucketJson(Newtonsoft.Json.JsonWriter,System.String,System.String)] Missing <param name="bucketType"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakPhaseInput.WriteBucketKeyBucketJson(Newtonsoft.Json.JsonWriter,System.String,System.String)] Missing <param name="writer"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.MapReduce.Inputs.RiakPhaseInput.WriteBucketKeyBucketJson(Newtonsoft.Json.JsonWriter,System.String,System.String)] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.Models.Search.Term.#ctor(RiakClient.Models.Search.RiakFluentSearch,System.String)] Missing <param name="field"/> documentation 
ShowMissingComponent: [M:RiakClient.Models.Search.Term.#ctor(RiakClient.Models.Search.RiakFluentSearch,System.String)] Missing <param name="search"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakClient.PutSearchIndex(RiakClient.Models.Search.SearchIndex)] Missing <param name="searchIndex"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakClient.PutSearchSchema(RiakClient.Models.Search.SearchSchema)] Missing <param name="searchSchema"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakCluster.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <param name="onError"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakCluster.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <param name="retryAttempts"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakCluster.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <param name="useFun"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakCluster.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <returns> documentation 
ShowMissingComponent: [M:RiakClient.RiakCluster.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.RiakCluster.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <typeparam name="TRiakResult"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakEndPoint.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <param name="onError"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakEndPoint.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <param name="retryAttempts"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakEndPoint.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <param name="useFun"/> documentation 
ShowMissingComponent: [M:RiakClient.RiakEndPoint.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <returns> documentation 
ShowMissingComponent: [M:RiakClient.RiakEndPoint.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <summary> documentation 
ShowMissingComponent: [M:RiakClient.RiakEndPoint.UseConnection``1(System.Func{RiakClient.Comms.IRiakConnection,``0},System.Func{RiakClient.ResultCode,System.String,System.Boolean,``0},System.Int32)] Missing <typeparam name="TRiakResult"/> documentation 
ShowMissingComponent: [P:RiakClient.Models.Index.BinIndex.IndexSuffix] Missing <summary> documentation 
ShowMissingComponent: [P:RiakClient.Models.Index.BinIndex.TypedThis] Missing <summary> documentation 
ShowMissingComponent: [P:RiakClient.Models.Index.IntIndex.IndexSuffix] Missing <summary> documentation 
ShowMissingComponent: [P:RiakClient.Models.Index.IntIndex.TypedThis] Missing <summary> documentation 
ShowMissingComponent: [P:RiakClient.Models.Index.SecondaryIndex`2.IndexSuffix] Missing <summary> documentation 
ShowMissingComponent: [P:RiakClient.Models.Index.SecondaryIndex`2.TypedThis] Missing <summary> documentation 
```

File: SECURITY.md [hunk 4]
```diff
@@ -83,13 +83,13 @@ Riak Security:
     <nodes>
       <node name="node1"
             hostAddress="riak-node1.mydomain.internal"
-            pbcPort="8097" restScheme="http" restPort="8098" poolSize="16" />
```
Comment (lukebakken) [inline SECURITY.md]: Good catch.

File: src/RiakClient.shfbproj [hunk 5]
```diff
@@ -30,38 +30,36 @@
     <SdkLinkTarget>Blank</SdkLinkTarget>
     <RootNamespaceContainer>True</RootNamespaceContainer>
     <PresentationStyle>VS2013</PresentationStyle>
-    <Preliminary>True</Preliminary>
+    <Preliminary>False</Preliminary>
```
Comment (lukebakken) [inline src/RiakClient.shfbproj]: awwyeah

File: src/RiakClient/Models/Index/RiakStreamedIndexResult.cs [hunk 6]
```diff
@@ -24,17 +24,21 @@ namespace RiakClient.Models.Index
...
+    /// <summary>
+    /// Represents a result to a streaming index query.
+    /// </summary>
     public class RiakStreamedIndexResult : IRiakIndexResult
     {
         private readonly IEnumerable<RiakResult<RpbIndexResp>> responseReader;
         private readonly bool includeTerms;
 
-        public RiakStreamedIndexResult(bool includeTerms, IEnumerable<RiakResult<RpbIndexResp>> responseReader)
+        internal RiakStreamedIndexResult(bool includeTerms, IEnumerable<RiakResult<RpbIndexResp>> responseReader)
```
Comment (lukebakken) [inline src/RiakClient/Models/Index/RiakStreamedIndexResult.cs]: This needs to be public as doing a `StreamGetSecondaryIndex` returns `RiakResult<RiakStreamedIndexResult>`. This applies to other `Riak...Result` classes too I think.

Comment (alexmoore) [inline src/RiakClient/Models/Index/RiakStreamedIndexResult.cs]: It shouldn't need to be public, the only thing that constructs the results is the RiakClient.cs class, the rest of the interface is public for consumption. (Unless I'm missing something).

Comment (lukebakken) [inline src/RiakClient/Models/Index/RiakStreamedIndexResult.cs]: Aha I missed that it was just the ctor. Can the ctors of the other result classes be changed to `internal` too?

Does making this `internal` affect someone who'd like to mock out things when using our lib?

File: src/RiakClient/Models/RiakCounterResult.cs [hunk 7]
```diff
@@ -19,25 +19,34 @@
...
 {
+    /// <summary>
+    /// Represents the data returned from a counter operation.
+    /// </summary>
     public class RiakCounterResult
     {
         private readonly RiakResult<RiakObject> result;
 
-        public RiakCounterResult(RiakResult<RiakObject> result)
+        internal RiakCounterResult(RiakResult<RiakObject> result)
```
Comment (lukebakken) [inline src/RiakClient/Models/RiakCounterResult.cs]: I have a gut feeling that public classes should have public ctors.

File: src/RiakClient/Models/Search/SearchIndex.cs [hunk 8]
```diff
@@ -23,35 +23,55 @@ namespace RiakClient.Models.Search
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SearchIndex"/> class.
+        /// </summary>
+        /// <param name="indexName">The name of the index.</param>
+        /// <param name="schemaName">The name of the schema for the index.</param>
+        /// <param name="nval">The <see cref="NVal"/> value for storing index entries.</param>
+        /// <exception cref="ArgumentException"><paramref name="indexName"/> cannot be null, zero length, or whitespace</exception>
+        /// <exception cref="ArgumentException"><paramref name="schemaName"/> cannot be null, zero length, or whitespace</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The value of <paramref name="nval"/> must not be null.</exception>
+        public SearchIndex(string indexName, string schemaName, NVal nval)
```
Comment (lukebakken) [inline src/RiakClient/Models/Search/SearchIndex.cs]: ty for improving argument names

**Comment by lukebakken** [comment]
I do think public classes should have public ctors. I have run into and have read about other people running into the situation where you really want to mock some class or test it or whatever but can't because there are no public ctors.

**Pull request merged**